drivers/sensor/lm35/lm35.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT lm35
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#define LM35_GAIN       ADC_GAIN_1
+#define LM35_REF        ADC_REF_INTERNAL
+#define LM35_RESOLUTION 12
+
+LOG_MODULE_REGISTER(LM35, CONFIG_SENSOR_LOG_LEVEL);
+
+struct lm35_data {
+	struct adc_channel_cfg ch_cfg;
+	uint16_t raw;
+};
+
+struct lm35_config {
+	const struct device *adc;
+	uint8_t adc_channel;
+	struct adc_sequence adc_seq;
+};
+
+static int lm35_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct lm35_config *cfg = dev->config;
+
+	return adc_read(cfg->adc, &cfg->adc_seq);
+}
+
+static int lm35_channel_get(const struct device *dev, enum sensor_channel chan,
+			    struct sensor_value *val)
+{
+	int err;
+	struct lm35_data *data = dev->data;
+	const struct lm35_config *cfg = dev->config;
+	int32_t mv = data->raw;
+
+	err = adc_raw_to_millivolts(adc_ref_internal(cfg->adc), data->ch_cfg.gain,
+				    cfg->adc_seq.resolution, &mv);
+	if (err) {
+		return err;
+	}
+
+	val->val1 = (int32_t)mv / 10;
+	val->val2 = (int32_t)mv % 10;
+
+	return 0;
+}
+
+static const struct sensor_driver_api lm35_driver_api = {
+	.sample_fetch = lm35_sample_fetch,
+	.channel_get = lm35_channel_get,
+};
+
+static int lm35_init(const struct device *dev)
+{
+	struct lm35_data *data = dev->data;
+	const struct lm35_config *cfg = dev->config;
+
+	if (!device_is_ready(cfg->adc)) {
+		LOG_ERR("ADC device is not ready.");
+		return -EINVAL;
+	}
+
+	data->ch_cfg = (struct adc_channel_cfg){
+		.gain = LM35_GAIN,
+		.reference = LM35_REF,
+		.acquisition_time = ADC_ACQ_TIME_DEFAULT,
+		.channel_id = cfg->adc_channel,
+	};
+
+	adc_channel_setup(cfg->adc, &data->ch_cfg);
+	return 0;
+}
+
+#define LM35_INST(inst)                                                                            \
+	static struct lm35_data lm35_data_##inst;                                                  \
+                                                                                                   \
+	static const struct lm35_config lm35_cfg_##inst = {                                        \
+		.adc = DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(inst)),                              \
+		.adc_channel = DT_INST_IO_CHANNELS_INPUT(inst),                                    \
+		.adc_seq = {                                                                       \
+			.channels = BIT(DT_INST_IO_CHANNELS_INPUT(inst)),                          \
+			.buffer = &lm35_data_##inst.raw,                                           \
+			.buffer_size = sizeof(lm35_data_##inst.raw),                               \
+			.resolution = LM35_RESOLUTION,                                             \
```

Comment:
What happens if the ADC driver doesn't support the preferred resolution? Will it re-scale? Or does an error occur?

Reply:
I think this depends on the ADC driver. From what I've seen, `adc_stm32.c` won't error out, while `adc_nrfx_adc.c` will return `-EINVAL`. esp32 will return `-ENOTSUP`.

I set it to 12 bits since other drivers at `/drivers/sensor` use this resolution, such as `grove/temperature_sensor.c` and `rpi_pico_temp/rpi_pico_temp.c` (also because my board supports this resolution).

Do you think we should have a different approach here? If so, how do you think we should proceed here?

Reply:
I looked at a few ADC drivers as well and made the same observations as you. I think the DT property is the best route for now. Another reviewer might have some more thoughts; I am personally not too familiar with the ADC API.

Longer term, I think it could be a good improvement to the ADC API to add a special value(s) like `ADC_RESOLUTION_MAX` or `ADC_RESOLUTION_DEFAULT` that all ADC drivers recognize to use the highest supported or defer to the ADC's own DT node if set. Applications like this sensor driver don't need necessarily need a specific ADC resolution and it'd be nice to have the option to abstract that away as long as `adc_raw_to_millivolts` does the right thing. I'll write up an issue for this.

Reply:
> I looked at a few ADC drivers as well and made the same observations as you. I think the DT property is the best route for now. Another reviewer might have some more thoughts; I am personally not too familiar with the ADC API.

Allright, changed to reflect the DT property. I've left it as optional defaulting to 12 bits.

> Longer term, I think it could be a good improvement to the ADC API to add a special value(s) like ADC_RESOLUTION_MAX or ADC_RESOLUTION_DEFAULT that all ADC drivers recognize to use the highest supported or defer to the ADC's own DT node if set. Applications like this sensor driver don't need necessarily need a specific ADC resolution and it'd be nice to have the option to abstract that away as long as adc_raw_to_millivolts does the right thing. I'll write up an issue for this.

Yes, I agree. From the application perspective, configuring ADC shouldn't be a concern.